### ISSUE TITLE
How to handle uniqueness constraints?

### ISSUE THREAD

**Issue by aflatter** [rom-rb/rom#101]
<a href="https://github.com/mcls"><img src="https://avatars.githubusercontent.com/u/714734?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [mcls](https://github.com/mcls)**
_Sunday Jan 11, 2015 at 11:05 GMT_
_Originally opened as https://github.com/rom-rb/rom-rails/issues/14_

----

I saw `validates_uniqueness_of` doesn't work.
I'm guessing this is because it requires access to the database during validation?

So I was wondering how (or where) these kind of "database-dependent" validations should be implemented?

**Comment by aflatter** [comment]
<a href="https://github.com/solnic"><img src="https://avatars.githubusercontent.com/u/1066?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [solnic](https://github.com/solnic)**
_Sunday Jan 11, 2015 at 11:32 GMT_

----

That's why you have separate validator objects. It's not an optimal solution but I've found it to be a good fit in Rails context. Here's how you could do it *now*:

``` ruby
class NewUserValidator
  def self.call(params)
    params.valid?
    params.errors.add(:email, :taken) unless unique_email?(params.email)
    params
  end

  def self.unique_email?(email)
    ROM.env.relations.users.where(email: email).count.zero?
  end
end
```

Please note that this is very ugly and verbose. I'm planning to add an interface for that so it will work without the need to write any custom code.

**Comment by aflatter** [comment]
<a href="https://github.com/mcls"><img src="https://avatars.githubusercontent.com/u/714734?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mcls](https://github.com/mcls)**
_Sunday Jan 11, 2015 at 11:45 GMT_

----

Ok, that makes sense :smiley:  

Now I'm confused about the `ActiveModel::Validations` being defined on the params/input object instead of the validator.

Is that because it's easier to include the `ActiveModel::Validations` module there?

**Comment by aflatter** [comment]
<a href="https://github.com/solnic"><img src="https://avatars.githubusercontent.com/u/1066?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [solnic](https://github.com/solnic)**
_Sunday Jan 11, 2015 at 11:50 GMT_

----

The reason for that is to meet Rails expectations which as an object with attributes *and* errors :( This way it Just Works™ with form helpers and libraries like simple_form.

I agree, it is confusing. I would love to find a better solution to this problem that would also work with Rails assumptions. I believe it should be possible to mix `ActiveModel::Validations` into validator module and do some tricks to make it work with params injected into a validator instance. I haven't tried to do it yet so I'm just guessing :)

Please notice that we want to separate input params and validations eventually, so this would actually be inline with what we want to achieve.

**Comment by aflatter** [comment]
<a href="https://github.com/mcls"><img src="https://avatars.githubusercontent.com/u/714734?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mcls](https://github.com/mcls)**
_Sunday Jan 11, 2015 at 12:15 GMT_

----

Ok, so I got this to work after accounting for a subtle issue.

In rails `#valid?` clear the errors first (see https://github.com/rails/rails/blob/v4.1.7.1/activemodel/lib/active_model/validations.rb#L314-L320). 

So in order to add errors in the validator object I had to check for `params.errors.blank?` instead of [`params.valid?`](https://github.com/rom-rb/rom-rails/blob/master/lib/rom/model.rb#L92) and then raise the `ROM::Model::ValidationError`.

Basically the snippet became:

```ruby
class NewUserValidator
  def self.call(params)
    params.valid?
    params.errors.add(:email, :taken) unless unique_email?(params.email)
    unless params.errors.blank?
      raise ROM::Model::ValidationError.new(params)
    end
  end

  def self.unique_email?(email)
    ROM.env.relations.users.where(email: email).count.zero?
  end
end
```

**Comment by aflatter** [comment]
<a href="https://github.com/solnic"><img src="https://avatars.githubusercontent.com/u/1066?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [solnic](https://github.com/solnic)**
_Sunday Jan 11, 2015 at 12:18 GMT_

----

@mcls oh sorry, totally forgot about raising the error part :) I reported a related issue which describes potential improvements => #15

**Comment by solnic** [comment]
FYI I started working on this and in [this branch](https://github.com/rom-rb/rom-rails/tree/improved-validators) it is already possible to do:

``` ruby
class UserParams
  include ROM::Model::Params

  attribute :name, String
  attribute :email, String
end

class UserValidator
  include ROM::Model::Validator

  relation :users

  validates :name, presence: true
  validates :email, uniqueness: true
end
```

It is now limited to only work with rom-sql and uses global ROM env but it's fine since we're within rails context exclusively here.

To make this work with all adapters we would have to come up with a common interface for validation-related relations. Something like `Relation#count_by(:attr_name)` or something.

**Comment by solnic** [comment]
This is now done and is supported by rom-sql and rom-rails. It'll be available in 0.6.0 of ROM and rails/sql updates.
